nsc-events-fullstack_29_122_e2e-playwright-testing

NahomAlemu opened the pull request:
## Summary & Changes 📃
- **Resolves:** #122

- **Summary:** Add comprehensive E2E testing infrastructure using Playwright
    - 🔨 Adds automated end-to-end testing to catch regressions before deployment
    - 👀 Tests should pass across Chromium, Firefox, WebKit, and Mobile Chrome
    - 🗨️ Tests run against both frontend (port 8080) and backend (port 3000) services

- **Changes:**
    - ✅ Added Playwright configuration with 4 browser projects
    - ✅ Added E2E tests for authentication (login, logout, password reset)
    - ✅ Added E2E tests for event management (create, edit, archive, unarchive)
    - ✅ Added E2E tests for event registration flow
    - ✅ Added E2E tests for user management and admin functions
    - ✅ Added accessibility tests (navigation, form labels, buttons, document title)
    - ✅ Added GitHub Actions workflow for CI/CD integration
    - ✅ Added test data generators and API client utilities
    - 🛠️ No breaking changes
    - 📝 Some tests skipped on Mobile Chrome due to MUI DatePicker limitations
    ⚠️ Known Issue: Event registration tests show 404 on some pages - will iterate on this in future updates


## Screenshots / Visual Aids 🔎

<details>
  <summary> Expand ⬇️ </summary>
  

https://github.com/user-attachments/assets/7844f8ad-55de-4942-a66f-b90dfe114680


</details>


## How to Test 🧪
1. **Steps to Reproduce:**
    - Step 1: Run `npm run test:e2e` (starts both services automatically)
    - Step 2: Or run `npm run test:e2e:ui` for interactive Playwright UI
    - Step 3: View test report: `npx playwright show-report` → http://localhost:9323
2. **Expected Behavior:** 107 tests pass, 5 skipped
3. **Actual Behavior (if bug):** All tests pass across 4 browser configurations


## Checklist ✅

- [X] I have **tested** this PR **locally** and it works as expected.
- [X] This PR **resolves an issue** (`Resolves #issue-number`).
- [X] **Reviewers, assignees(self), tags, and labels** are correctly assigned. <!-- on the menu to the right -->
- [ ] **Squash commits** and enable **auto-merge** if approved.

NahomAlemu committed:
Resolve package.json conflict and reapply stashed changes

NahomAlemu committed:
Add E2E testing with Playwright

NahomAlemu committed:
Install sharp for Linux platform in backend CI

NahomAlemu committed:
Rebuild sharp at monorepo root for Linux

NahomAlemu committed:
Use npm rebuild for sharp platform fix

NahomAlemu committed:
Upgrade Playwright from 1.42.1 to 1.57.0

gitbiruk2010 committed:
Merge branch 'main' into feature-122-e2e-playwright-testing

NahomAlemu committed:
add missing sharp rebuild

gitbiruk2010 committed:
ci: fix PR and e2e workflows for Sharp build

- Add --include=optional to
pm ci in on-pull-request and e2e workflows
- Install libvips-dev before running
pm rebuild sharp to prevent Sharp build failures

gitbiruk2010 committed:
fix: workflow update to use - name: Install sharp for Linux platform
  run: npm install --os=linux --cpu=x64 sharp
  working-directory: ./nsc-events-nestjs. this will explicitly installs the Linux x64 prebuilt binaries for sharp directly in the nsc-events-nestjs workspace

gitbiruk2010 committed:
fix: workflow e2e-test.yml - Changed PostgreSQL service: POSTGRES_DB: nsc_events_test → POSTGRES_DB: nsc_events && Replaced DATABASE_URL with individual env vars that the backend actually reads

gitbiruk2010 committed:
Fix: Added JWT_SECRET: e2e-test-jwt-secret-key to the environment variables in the Run E2E tests step of .github/workflows/e2e-tests.yml.

gitbiruk2010 committed:
add cache playwright browsers in e2e workflow

gitbiruk2010 committed:
fix: resolve E2E test failures in CI environment - Pass environment variables to Playwright webServer backend process
- Skip .env file loading when NODE_ENV=test (use process.env instead)
- Add explicit validation for missing JWT_SECRET at startup

gitbiruk2010 commented:
@NahomAlemu  i tried to figure the issue and have made some progress. 
The backend ci is not failing now. and results are visible while some tests are still failing. feel free to take over. We will see if we can fix it by the end of the week. if not, unfortunately we will need to backlog this for the next quarter for someone to take over. 

**From:**
<img width="412" height="262" alt="image" src="https://github.com/user-attachments/assets/2be5a9bc-fcbd-44a5-a040-25513db18ff2" />

**To:**
<img width="452" height="257" alt="image" src="https://github.com/user-attachments/assets/36a304e5-b9b4-48ec-aa6e-0b635e3732e9" />

gitbiruk2010 committed:
fix: - Add /api suffix to PLAYWRIGHT_API_URL in workflow (fixes 404 errors)
- Pass JWT_SECRET and other env vars to Playwright webServer process

NahomAlemu committed:
fix test fails and add date-fns

NahomAlemu committed:
Fix E2E workflow: correctly build NestJS backend before tests

NahomAlemu committed:
resolve missing build:backend script in e2e workflow

NahomAlemu commented:
> @NahomAlemu i tried to figure the issue and have made some progress. The backend ci is not failing now. and results are visible while some tests are still failing. feel free to take over. We will see if we can fix it by the end of the week. if not, unfortunately we will need to backlog this for the next quarter for someone to take over.

Thank you for your help on this! I wasn't able to figure out the root cause yet. Will work on it later. Converting it to draft...

NahomAlemu committed:
correct monorepo build steps for backend and frontend

NahomAlemu committed:
add playwright caching back

gitbiruk2010 committed:
test: add comprehensive unit tests for queries utility and fix stale data bug- Added unit tests for React Query hooks with full coverage- Fixed useMyEvents query key missing userId causing stale cache between users- Tested all hooks: useFilteredEvents, useMyEvents, useEventById, useArchivedEvents, useIsAttending- Covers query configuration, API URL construction, data transformation, error handling (#126)

gitbiruk2010 committed:
test: add unit tests for dbFieldMapper and fix falsy id bug (#128)

- Add full cov test cases for dbFieldMapper utility
- Fix bug in getActivityId: change || to ?? (nullish coalescing)
- Test normalizeActivityId, normalizeActivityIds, and getActivityId
- Cover null/undefined handling, field mapping, and data preservation
- Test edge cases including empty strings, zero values, and nested objects

gitbiruk2010 committed:
feat: add unregister confirmation dialog, update HTTPS link, and cleanup test files (#129)

* test: add unit tests for dbFieldMapper and fix falsy id bug

- Add full cov test cases for dbFieldMapper utility
- Fix bug in getActivityId: change || to ?? (nullish coalescing)
- Test normalizeActivityId, normalizeActivityIds, and getActivityId
- Cover null/undefined handling, field mapping, and data preservation
- Test edge cases including empty strings, zero values, and nested objects

* feat: add unregister confirmation dialog, fix HTTPS link, remove test files

- Add confirmation dialog for unregister button on event detail page
- Update Central Seattle link in Footer from HTTP to HTTPS
- Remove unused test files: RemoveTest.tsx, TEST.md, test.localhook.md

NahomAlemu committed:
Fix start:prod script path to dist/src/main

NahomAlemu committed:
Fix: Add NEXT_PUBLIC_API_URL for E2E tests in CI

beimnettes reviewed (approved):
I pulled the branch locally and verified the Playwright setup. Tests are discovered correctly across multiple browsers, and the coverage looks solid for auth, events, registration, admin flows, and accessibility. I also followed the CI updates and it’s great to see the full test suite passing. This is a strong addition to the project.
<img width="1023" height="110" alt="Screenshot 2026-01-22 at 7 00 28 PM" src="https://github.com/user-attachments/assets/bd9b891c-9f2d-4493-8f60-152df03f5936" />

IsaacJrTypes committed:
Add db dump file for schema and data population

IsaacJrTypes committed:
Remove db dump file

IsaacJrTypes reviewed (approved):
I was able to get the tests up and running! Thanks for the tips Nahom. I removed the dump file, but I wanted to save a copy for future reference.
